File: services/tools/fs.py
from __future__ import annotations

import re
from pathlib import Path

from services.tools.registry import register_tool
from services.tools.workspace import (
    read_text_limited,
    resolve_path,
    tool_error,
    tool_ok,
    workspace_root,
)

MAX_LIST_ENTRIES = 500
MAX_GLOB_MATCHES = 200
MAX_GREP_MATCHES = 100
MAX_GREP_FILE_BYTES = 500_000


def _rel(path: Path) -> str:
    return str(path.relative_to(workspace_root()))


@register_tool()
def read_file(path: str) -> str:
    """读取工作区内的文本文件。path 为相对工作区的路径。"""
    try:
        resolved = resolve_path(path, must_exist=True)
        if not resolved.is_file():
            return tool_error(f"不是文件: {resolved}")
        content = read_text_limited(resolved)
        return tool_ok(path=_rel(resolved), content=content)
    except Exception as exc:
        return tool_error(str(exc))


@register_tool()
def list_dir(path: str = ".") -> str:
    """列出工作区目录下的文件和子目录。path 默认为工作区根目录。"""
    try:
        resolved = resolve_path(path, must_exist=True)
        if not resolved.is_dir():
            return tool_error(f"不是目录: {resolved}")
        entries = sorted(resolved.iterdir(), key=lambda p: (not p.is_dir(), p.name.lower()))
        truncated = len(entries) > MAX_LIST_ENTRIES
        if truncated:
            entries = entries[:MAX_LIST_ENTRIES]
        items = [
            {
                "name": p.name,
                "type": "dir" if p.is_dir() else "file",
                "path": _rel(p),
            }
            for p in entries
        ]
        return tool_ok(
            path=_rel(resolved),
            entries=items,
            truncated=truncated,
            total=len(items),
        )
    except Exception as exc:
        return tool_error(str(exc))


@register_tool()
def search_files(pattern: str, path: str = ".") -> str:
    """按 glob 模式搜索文件，例如 **/*.json 或 src/**/*.tsx。"""
    try:
        base = resolve_path(path, must_exist=True)
        if not base.is_dir():
            return tool_error(f"不是目录: {base}")
        all_matches = sorted(
            (p for p in base.glob(pattern) if p.is_file()),
            key=lambda p: str(p),
        )
        truncated = len(all_matches) > MAX_GLOB_MATCHES
        matches = [_rel(p) for p in all_matches[:MAX_GLOB_MATCHES]]
        return tool_ok(
            pattern=pattern,
            path=_rel(base),
            matches=matches,
            truncated=truncated,
            total=len(all_matches),
        )
    except Exception as exc:
        return tool_error(str(exc))


@register_tool()
def grep_files(
    pattern: str,
    path: str = ".",
    glob: str = "**/*",
) -> str:
    """在文件内容中搜索正则 pattern，返回匹配行。glob 限定文件范围，默认 **/*。"""
    try:
        base = resolve_path(path, must_exist=True)
        if not base.is_dir():
            return tool_error(f"不是目录: {base}")
        regex = re.compile(pattern)
        hits: list[dict[str, str | int]] = []
        for file_path in sorted(base.glob(glob)):
            if not file_path.is_file():
                continue
            if file_path.stat().st_size > MAX_GREP_FILE_BYTES:
                continue
            try:
                text = file_path.read_text(encoding="utf-8", errors="replace")
            except OSError:
                continue
            for line_no, line in enumerate(text.splitlines(), start=1):
                if regex.search(line):
                    hits.append({"path": _rel(file_path), "line": line_no, "text": line[:300]})
                    if len(hits) >= MAX_GREP_MATCHES:
                        return tool_ok(
                            pattern=pattern,
                            path=_rel(base),
                            matches=hits,
                            truncated=True,
                        )
        return tool_ok(pattern=pattern, path=_rel(base), matches=hits, truncated=False)
    except re.error as exc:
        return tool_error(f"无效正则: {exc}")
    except Exception as exc:
        return tool_error(str(exc))


@register_tool()
def write_file(path: str, content: str) -> str:
    """写入工作区内的文本文件。仅在用户明确要求修改时使用。"""
    try:
        resolved = resolve_path(path)
        resolved.parent.mkdir(parents=True, exist_ok=True)
        resolved.write_text(content, encoding="utf-8")
        return tool_ok(path=_rel(resolved), bytes=len(content.encode("utf-8")))
    except Exception as exc:
        return tool_error(str(exc))

File: apps/server/src/tools/fs.ts
import fs from "node:fs/promises";
import path from "node:path";

import { registerTool } from "./registry.js";
import {
  readTextLimited,
  relPath,
  resolvePath,
  toolError,
  toolOk,
  workspaceRoot,
} from "./workspace.js";

const MAX_LIST_ENTRIES = 500;
const MAX_GLOB_MATCHES = 200;
const MAX_GREP_MATCHES = 100;
const MAX_GREP_FILE_BYTES = 500_000;

async function walkFiles(base: string, pattern: string): Promise<string[]> {
  const glob = pattern.includes("*") ? pattern : `**/${pattern}`;
  const results: string[] = [];

  async function walk(dir: string): Promise<void> {
    const entries = await fs.readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        await walk(full);
      } else if (entry.isFile()) {
        results.push(full);
      }
    }
  }

  await walk(base);

  const minimatch = (file: string, pat: string): boolean => {
    const rel = path.relative(base, file).replace(/\\/g, "/");
    const re = new RegExp(
      "^" +
        pat
          .replace(/\*\*/g, "§§")
          .replace(/\*/g, "[^/]*")
          .replace(/§§/g, ".*")
          .replace(/\./g, "\\.") +
        "$",
    );
    return re.test(rel) || rel.endsWith(pat.replace(/\*/g, ""));
  };

  return results.filter((f) => minimatch(f, glob)).slice(0, MAX_GLOB_MATCHES + 1);
}

registerTool({
  name: "read_file",
  description: "读取工作区内的文本文件。path 为相对工作区的路径。",
  parameters: {
    type: "object",
    properties: { path: { type: "string" } },
    required: ["path"],
  },
  handler: async (args) => {
    try {
      const resolved = resolvePath(String(args.path), true);
      const stat = await fs.stat(resolved);
      if (!stat.isFile()) return toolError(`不是文件: ${resolved}`);
      const content = await readTextLimited(resolved);
      return toolOk({ path: relPath(resolved), content });
    } catch (err) {
      return toolError(err instanceof Error ? err.message : String(err));
    }
  },
});

registerTool({
  name: "list_dir",
  description: "列出工作区目录下的文件和子目录。path 默认为工作区根目录。",
  parameters: {
    type: "object",
    properties: { path: { type: "string" } },
  },
  handler: async (args) => {
    try {
      const resolved = resolvePath(String(args.path ?? "."), true);
      const stat = await fs.stat(resolved);
      if (!stat.isDirectory()) return toolError(`不是目录: ${resolved}`);
      const entries = await fs.readdir(resolved, { withFileTypes: true });
      entries.sort((a, b) => {
        if (a.isDirectory() !== b.isDirectory()) return a.isDirectory() ? -1 : 1;
        return a.name.localeCompare(b.name);
      });
      const truncated = entries.length > MAX_LIST_ENTRIES;
      const slice = truncated ? entries.slice(0, MAX_LIST_ENTRIES) : entries;
      const items = await Promise.all(
        slice.map(async (e) => ({
          name: e.name,
          type: e.isDirectory() ? "dir" : "file",
          path: relPath(path.join(resolved, e.name)),
        })),
      );
      return toolOk({
        path: relPath(resolved),
        entries: items,
        truncated,
        total: items.length,
      });
    } catch (err) {
      return toolError(err instanceof Error ? err.message : String(err));
    }
  },
});

registerTool({
  name: "search_files",
  description: "按 glob 模式搜索文件，例如 **/*.json 或 src/**/*.tsx。",
  parameters: {
    type: "object",
    properties: {
      pattern: { type: "string" },
      path: { type: "string" },
    },
    required: ["pattern"],
  },
  handler: async (args) => {
    try {
      const base = resolvePath(String(args.path ?? "."), true);
      const stat = await fs.stat(base);
      if (!stat.isDirectory()) return toolError(`不是目录: ${base}`);
      const pattern = String(args.pattern);
      const all = await walkFiles(base, pattern);
      const truncated = all.length > MAX_GLOB_MATCHES;
      const matches = all.slice(0, MAX_GLOB_MATCHES).map(relPath);
      return toolOk({
        pattern,
        path: relPath(base),
        matches,
        truncated,
        total: all.length,
      });
    } catch (err) {
      return toolError(err instanceof Error ? err.message : String(err));
    }
  },
});

registerTool({
  name: "grep_files",
  description: "在文件内容中搜索正则 pattern，返回匹配行。glob 限定文件范围，默认 **/*。",
  parameters: {
    type: "object",
    properties: {
      pattern: { type: "string" },
      path: { type: "string" },
      glob: { type: "string" },
    },
    required: ["pattern"],
  },
  handler: async (args) => {
    try {
      const base = resolvePath(String(args.path ?? "."), true);
      const globPat = String(args.glob ?? "**/*");
      const regex = new RegExp(String(args.pattern));
      const files = await walkFiles(base, globPat);
      const hits: Array<{ path: string; line: number; text: string }> = [];
      for (const file of files) {
        const st = await fs.stat(file);
        if (st.size > MAX_GREP_FILE_BYTES) continue;
        const text = await fs.readFile(file, "utf8");
        for (const [i, line] of text.split("\n").entries()) {
          if (regex.test(line)) {
            hits.push({ path: relPath(file), line: i + 1, text: line.slice(0, 300) });
            if (hits.length >= MAX_GREP_MATCHES) {
              return toolOk({
                pattern: String(args.pattern),
                path: relPath(base),
                matches: hits,
                truncated: true,
              });
            }
          }
        }
      }
      return toolOk({
        pattern: String(args.pattern),
        path: relPath(base),
        matches: hits,
        truncated: false,
      });
    } catch (err) {
      return toolError(err instanceof Error ? err.message : String(err));
    }
  },
});

registerTool({
  name: "write_file",
  description: "写入工作区内的文本文件。仅在用户明确要求修改时使用。",
  parameters: {
    type: "object",
    properties: {
      path: { type: "string" },
      content: { type: "string" },
    },
    required: ["path", "content"],
  },
  handler: async (args) => {
    try {
      const resolved = resolvePath(String(args.path));
      await fs.mkdir(path.dirname(resolved), { recursive: true });
      const content = String(args.content);
      await fs.writeFile(resolved, content, "utf8");
      return toolOk({
        path: relPath(resolved),
        bytes: Buffer.byteLength(content, "utf8"),
      });
    } catch (err) {
      return toolError(err instanceof Error ? err.message : String(err));
    }
  },
});

// side-effect: ensure workspace root exists
void workspaceRoot();
